### PR TITLE
Use a string compatible prefix comparator

### PR DESCRIPTION
**Pull request by jeancroy**
fix https://github.com/atom/autocomplete-snippets/issues/61.
May also help [related](https://github.com/atom/language-ruby/issues/130) [issues](https://github.com/jeancroy/fuzzaldrin-plus/issues/9)

### CODE REVIEW

**Comment by jeancroy** [comment]
/cc @50Wliu because I think this would help the ruby do/dop snippet situation.

I'm not sure if spec are needed or how I'd write one.

**Comment by lee-dohm** [comment]
@jeancroy Specs are strongly preferred. Basically I would just write a bunch of tests describing sorting cases and the expected output. Extra credit for covering failure cases like people handing `undefined`, `null` or non-string input.

**Comment by jasonewall** [comment]
@lee-dohm How are the tests for this repo executed?

**Comment by jeancroy** [comment]
@lee-dohm would it be OK to test the sortComparator in isolation ?
I can give an array of unsorted data, and try to sort using this.

it's just the inner working of autocomplete-snippet i'm not sure. It looks like a glue package between the snippet package and autocomplete-plus.

**Comment by lee-dohm** [comment]
Yes, that's what I was thinking.

**Comment by jeancroy** [comment]
However i think it's exist only as an utility function available in the closure. I'm not sure I can reference to it from the ouput of a require.

**Comment by jeancroy** [comment]
Thanks @lee-dohm, I blindly assumed at least the a.prefix part was right. It looks like result don't have a prefix property but  a text one.

**Comment by lee-dohm** [comment]
You're welcome :+1:

File: spec/autocomplete-snippets-spec.coffee [hunk 1]
```diff
@@ -71,3 +71,27 @@ describe 'AutocompleteSnippets', ->
       runs ->
         atom.commands.dispatch(editorView, 'autocomplete-plus:confirm')
         expect(editor.getText()).toContain '} while (true);'
+
```
Comment (benogle) [inline spec/autocomplete-snippets-spec.coffee]: :fire: this newliine

File: spec/autocomplete-snippets-spec.coffee [hunk 2]
```diff
@@ -71,3 +71,27 @@ describe 'AutocompleteSnippets', ->
...
+
+  describe 'when showing suggestions', ->
+    it 'sort them in alphabetical order', ->
```
Comment (benogle) [inline spec/autocomplete-snippets-spec.coffee]: `sort` -> `sorts`

File: spec/autocomplete-snippets-spec.coffee [hunk 3]
```diff
@@ -71,3 +71,27 @@ describe 'AutocompleteSnippets', ->
...
+
```
Comment (benogle) [inline spec/autocomplete-snippets-spec.coffee]: :fire: this newline

File: lib/snippets-provider.coffee [hunk 4]
```diff
@@ -43,7 +43,10 @@ class SnippetsProvider
   onDidInsertSuggestion: ({editor}) ->
     atom.commands.dispatch(atom.views.getView(editor), 'snippets:expand')
 
-ascendingPrefixComparator = (a, b) -> a.prefix  - b.prefix
+
+# To the choice of reviewer
+ascendingPrefixComparator = (a, b) -> if a.text > b.text then 1 else if a.text < b.text then -1 else 0
+#ascendingPrefixComparator = (a, b) -> a.text.localeCompare(b.text)
```
Comment (benogle) [inline lib/snippets-provider.coffee]: I'm leaning toward the `localeCompare`. Trying to figure out the downsides of it.

Comment (jeancroy) [inline lib/snippets-provider.coffee]: Should we guard against null/undefined text ?
`a.text?.localeCompare(b.text)`

**Pull request merged**